lib/quicly.c

```diff
@@ -4483,15 +4774,216 @@ static int handle_ping_frame(quicly_conn_t *conn, struct st_quicly_handle_payloa
     return 0;
 }
 
+/**
+ * copies NEW_CONNECTION_ID information to an internal storage from a received frame
+ */
+static void store_spare_cid(struct st_quicly_spare_cid_t *spare_cid, const quicly_new_connection_id_frame_t *frame)
+{
+    spare_cid->sequence = frame->sequence;
+    set_cid(&spare_cid->cid, frame->cid);
+    memcpy(spare_cid->stateless_reset_token, frame->stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN);
+    spare_cid->is_active = 1;
+}
+
+/**
+ * compares CIDs, returns zero if identical
+ */
+static int cid_is_equal(const quicly_cid_t *l, ptls_iovec_t r)
+{
+    if (l->len != r.len)
+        return 1;
+    return memcmp(l->cid, r.base, l->len);
+}
+
+/**
+ * checks if the received CID carries conflicting information with CIDs we already have
+ *
+ * @param sequence known sequence number
+ * @param cid known CID
+ * @param stateless_reset_token known stateless reset token
+ * @param frame newly received NEW_CONNECTION_ID frame
+ * @param retcode return code that the caller will return to its caller
+ * @return 0 if okay (the caller can continue), non-zero otherwise (the caller must terminate frame rocessing and return *retcode)
+ */
+static int verify_new_cid(uint64_t sequence, const quicly_cid_t *cid,
+                          const uint8_t stateless_reset_token[QUICLY_STATELESS_RESET_TOKEN_LEN],
+                          const quicly_new_connection_id_frame_t *frame, int *retcode)
```

Comment:
👍 on renaming arguments.

For the return code, it's not that trivial. In case of valid but duplicated frame, this function has to tell the caller "it's okay but stop execution". In other words, simply returning zero doesn't distinguish "okay and continue" and "okay but return immediately".

Reply:
Renaming done.

Reply:
Thank you for renaming the arguments. With the changes, I think it has become easier to read.

Re return code, reading your comment, I now realize that this function is used for two purposes: (1) verify if the provided values are correct, (2) signal to the caller that there's nothing to be done.

I think my concern now is that it's confusing, regarding the name and how it's used.

```c
if (verify_new_cid(conn->super.peer.cid_sequence, &conn->super.peer.cid, conn->super.peer.stateless_reset.token, &frame, &ret))
         return ret;
```

It's a bit difficult to understand that we are deviating from our ordinary way of doing `if ((ret = foo()) != 0) return ret`, and that in this particular case we might be returning zero.

How about dropping the (2)? I believe that can be done in the caller; i.e. `if (seq == frame.seq) return 0`.

Or, we could lift the implication that the function is merely verifying something, by changing the name to something like "check_new_cid" and do:
```c
if ((ret = check_new_cid(conn->super.peer.cid_sequence, &conn->super.peer.cid, &conn->super.peer.stateless_reset_token, &frame, &found)) != 0 || found)
    return ret;
```

Anyways, this is my slight preference, and it might be the case that you might not need to look into this now... we might change the way this function is implemented as we refactor the code.